src/endpoints/tps/tps.service.ts

```diff
@@ -0,0 +1,52 @@
+import { Injectable } from "@nestjs/common";
+import { TpsFrequency } from "./entities/tps.frequency";
+import { TpsUtils } from "src/utils/tps.utils";
+import { CacheService } from "@multiversx/sdk-nestjs-cache";
+import { CacheInfo } from "src/utils/cache.info";
+import { Tps } from "./entities/tps";
+import { TpsInterval } from "./entities/tps.interval";
+
+@Injectable()
+export class TpsService {
+  constructor(
+    private readonly cacheService: CacheService,
+  ) { }
+
+  async getTpsLive(frequency: TpsFrequency): Promise<Tps> {
+    const frequencySeconds = TpsUtils.getFrequencyByEnum(frequency);
+    const timestamp = TpsUtils.getTimestampByFrequency(new Date().getTimeInSeconds() - frequencySeconds, frequencySeconds);
+
+    const transactions = (await this.cacheService.getRemote<number>(CacheInfo.TpsByTimestampAndFrequency(timestamp, frequencySeconds).key)) ?? 0;
+
+    const tps = transactions / frequencySeconds;
+
+    return new Tps({ timestamp, tps });
+  }
+
+  async getTpsHistory(interval: TpsInterval): Promise<Tps[]> {
+    return await this.cacheService.getOrSet<Tps[]>(
+      CacheInfo.TpsHistoryByInterval(interval).key,
+      async () => await this.getTpsHistoryRaw(interval),
+      CacheInfo.TpsHistoryByInterval(interval).ttl
+    );
+  }
+
+  async getTpsHistoryRaw(interval: TpsInterval): Promise<Tps[]> {
+    const frequencySeconds = TpsUtils.getFrequencyByInterval(interval);
+    const endTimestamp = TpsUtils.getTimestampByFrequency(new Date().getTimeInSeconds() - frequencySeconds, frequencySeconds);
```

Comment:
if the timestamp is 37 and the frequency is 3, we'll get an end timestamp of:
```
TpsUtils.getTimestampByFrequency(37 - 3, 3); // 34, 3
```
which (considering the logic of the `getTimestampByFrequency` function) will result in:
```
endTimestamp = Math.floor(34 / 3) * 3; // = 33
```
but we should already have the tps computed for the entrypoint 36. 
expected?

Reply:
that is correct!
`getTimestampByFrequency` gets the beginning of the provided interval which should always be in the past or present. so we should have the data indexed up until that point
removed the subtraction of the frequency in the historical case. for the latest tps it makes sense to fetch the latest closed interval to make sure the data associated with finalized intervals is fetched